alphazero/mcts/mcts.py

```diff
@@ -0,0 +1,282 @@
+        """
+        if not rollout:
+            _, v_leaf = leaf_node.calc_policy_value(
+                rollout=rollout, network=net
+            )  # Think about batching/efficiency
+        else:
+            _, v_leaf = leaf_node.calc_policy_value(rollout=rollout, game=game)
+
+        leaf_node.is_terminal = game.is_game_over(leaf_node.state)
+        if not leaf_node.is_terminal:
+            leaf_node.expand(game)
+        current = leaf_node
+        while not current.is_root:
+            current.n_visit += 1
+            current.total_value += v_leaf
+            current = current.parent
+
+    def forward(root_node: MCTSNode):
+        """
+        Runs a forward pass through the tree and then backpropagates back through the tree
+        Args:
+            root_node (MCTSNode): Root node to start from
+        """
+        current = root_node
+        while not current.is_leaf and not current.is_terminal:
+            sum_n = np.sum([child.n_visit for child in current.children])
+            current = current.children[
+                np.argmax(
+                    [
+                        child.confidence_bound(sum_n=sum_n, c_puct=c_puct)
+                        for child in current.children
+                    ]
+                )
+            ]
+        backpropagate(current)
+
+    for _ in range(n_iter):
+        forward(start_node)
+    policy_count = torch.LongTensor([child.n_visit for child in start_node.children])
+    sampled_node = start_node.children[
+        torch.multinomial(torch.pow(policy_count, (1 / temperature)).float(), 1).item()
+    ]
+    # TODO Maybe just iterate through children once, more efficient?
+    # Currently doing it twice due to output difference, just do forloop instead?
+    return (
+        sum([child.action * child.n_visit for child in start_node.children]),
+        sampled_node,
+    )
+
+
+def self_play(
+    game: GameClassType, net: torch.nn.Module, n_mcts_iter: int = 1600,
+) -> List[tuple]:
+    pos = game.get_init_board()
+    temperature = 0.0001
+    data = []
+    is_terminal, res = game.is_terminal(pos)
+    node = MCTSNode(state=pos)
+    while not is_terminal:
+        policy, node = mcts(
+            start_node=node,
+            game=game,
+            net=net,
+            n_iter=n_mcts_iter,
+            temperature=temperature,
+        )
+        data.append((pos, policy))
+        pos = game.make_move(pos, node.action)
+        is_terminal, res = game.is_terminal(pos)
+        # TODO: Currently does not add last state (when game ends, no moves available) to datapoints, should it? Also look into stopping thresholds and resigning
+    data = [row + (res,) for row in data]
+
+    return data
+
+
+# def play_games(
+#    game: Game,
+#    net: torch.nn.Module,
+#    n_mcts_iter: int = 1600,
+#    n_games_thread=2500,
+#    threads=4,
+# ):
+#    # Arg games/thread instead?
+#    def play_games_thread():
+#        data = []
+#        for i in range(n_games_thread):
```

Comment:
can we remove this? maybe you can add it to your stash to keep it around?

Reply:
Sure, I was just lazy